agent/dockerclient/dockerapi/docker_client.go

```diff
@@ -1221,31 +1222,61 @@ func (dg *dockerGoClient) APIVersion() (dockerclient.DockerVersion, error) {
 	return dg.sdkClientFactory.FindClientAPIVersion(client), nil
 }
 
-// Stats returns a channel of *docker.Stats entries for the container.
-func (dg *dockerGoClient) Stats(id string, ctx context.Context) (<-chan *docker.Stats, error) {
-	client, err := dg.dockerClient()
+// Stats returns a channel of *types.Stats entries for the container.
+func (dg *dockerGoClient) Stats(id string, ctx context.Context) (<-chan *types.Stats, error) {
+	subCtx, cancelRequest := context.WithCancel(ctx)
+	defer cancelRequest()
+
+	client, err := dg.sdkDockerClient()
 	if err != nil {
 		return nil, err
 	}
 
-	stats := make(chan *docker.Stats)
-	options := docker.StatsOptions{
-		ID:                id,
-		Stats:             stats,
-		Stream:            true,
-		Context:           ctx,
-		InactivityTimeout: StatsInactivityTimeout,
-	}
+	// Create channel to hold the stats
+	statsChnl := make(chan *types.Stats)
+	var resp types.ContainerStats
 
 	go func() {
-		statsErr := client.Stats(options)
-		if statsErr != nil {
-			seelog.Infof("DockerGoClient: Unable to retrieve stats for container %s: %v",
-				id, statsErr)
+		// ContainerStats outputs a io.ReadCloser and an OSType
+		stream := true
+		resp, err = client.ContainerStats(subCtx, id, stream)
+		if err != nil {
+			seelog.Warnf("DockerGoClient: Unable to retrieve stats for container %s: %v", id, err)
+			close(statsChnl)
+			return
+		}
+
+		// handle inactivity timeout
+		var canceled uint32
+		var ch chan<- struct{}
+		resp.Body, ch = handleInactivityTimeout(resp.Body, StatsInactivityTimeout, cancelRequest, &canceled)
+		defer resp.Body.Close()
+		defer close(ch)
+
+		// Returns a *Decoder and takes in a readCloser
+		decoder := json.NewDecoder(resp.Body)
+		data := new(types.Stats)
+		for err := decoder.Decode(data); err != io.EOF; err = decoder.Decode(data) {
```

Comment:
Should we add an unit test covering error `io.EOF`?

Reply:
io.EOF is covered in the NormalExit test case.